### PR TITLE
Refactor the authorityIssuer and add specificIssuer support

### PR DESCRIPTION
**Pull request by chaoxinhu**
This change refactorized the authority issuer, committee member and add support of specific issuer list on chain.

### CODE REVIEW

File: contracts/SpecificIssuerData.sol [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+ * @title SpecificIssuerData
+ * Stores data about issuers with specific types.
+ */
+
+contract SpecificIssuerData {
+    struct IssuerType {
+        // typeName as index, dynamic array as getAt function and mapping as search
+        bytes32 typeName;
+        address[] member;
+        mapping (address => bool) isMember;
```
Comment (chaoxinhu) [inline contracts/SpecificIssuerData.sol]: add extra field

File: contracts/AuthorityIssuerController.sol [hunk 2]
```diff
@@ -89,16 +95,29 @@ contract AuthorityIssuerController {
         }
     }
 
-    function getAllAuthorityIssuerAddress() 
+    function getAllAuthorityIssuerAddress(
```
Comment (junqizhang-dev) [inline contracts/AuthorityIssuerController.sol]: Please rename this function to getAuthorityIssuerAddressList

File: contracts/AuthorityIssuerController.sol [hunk 3]
```diff
@@ -60,6 +62,10 @@ contract AuthorityIssuerController {
     )
         public
     {
+        if (authorityIssuerData.isNameDuplicate(attribBytes32[0])) {
```
Comment (junqizhang-dev) [inline contracts/AuthorityIssuerController.sol]: Please delete this if statement, and it only needs to be called once in data contract.

File: contracts/AuthorityIssuerData.sol [hunk 4]
```diff
@@ -72,7 +74,7 @@ contract AuthorityIssuerData {
     )
         public
     {
-        if (isAuthorityIssuer(addr)) {
+        if (isAuthorityIssuer(addr) || isNameDuplicate(attribBytes32[0])) {
```
Comment (junqizhang-dev) [inline contracts/AuthorityIssuerData.sol]: Please add return value for this function.

File: contracts/SpecificIssuerData.sol [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+    }
+
+    function isSpecificTypeIssuer(bytes32 typeName, address addr) public constant returns (bool) {
+        if (issuerTypeMap[typeName].isMember[addr] == false) {
+            return false;
+        }
+        return true;
+    }
+
+    function getSpecificTypeIssuerMembers(bytes32 typeName, uint startPos) public constant returns (address[50]) {
```
Comment (junqizhang-dev) [inline contracts/SpecificIssuerData.sol]: Please rename getSpecificTypeIssuerMembers to getSpecificTypeIssuerMemberList

File: contracts/SpecificIssuerData.sol (hunk 1 above)
Comment (chaoxinhu) [reply contracts/SpecificIssuerData.sol]: CRU of extra field done

File: contracts/AuthorityIssuerController.sol (hunk 2 above)
Comment (chaoxinhu) [reply contracts/AuthorityIssuerController.sol]: done

File: contracts/AuthorityIssuerController.sol (hunk 3 above)
Comment (chaoxinhu) [reply contracts/AuthorityIssuerController.sol]: fully refactored

File: contracts/AuthorityIssuerData.sol (hunk 4 above)
Comment (chaoxinhu) [reply contracts/AuthorityIssuerData.sol]: fully refactorized

**Review by junqizhang-dev — APPROVED** [review]

**Pull request merged**